software/firmware/europi.py

```diff
@@ -1,249 +1,278 @@
+    def set_samples(self, samples):
+        """Override the default number of sample reads with the given value."""
+        if not isinstance(samples, int):
+            raise ValueError(
+                f"set_samples expects an int value, got: {samples}")
+        self._samples = samples
 
-    def unit_interval(self, samples=256):
-        return 1 - (sample_adc(self.input, samples) / MAX_UINT12) #Provide the knob's position as a float value between 0.0 and 1.0
+    def percent(self, samples=None):
+        """Return the percentage of the component's current relative range."""
+        return self._sample_adc(samples) / MAX_UINT16
 
-    def read_position(self, steps=100, samples=256): #Returns an int in the range of steps based on knob position.
+    def range(self, steps=100, samples=None):
+        """Return a value from steps chosen by the current voltages relative position."""
         if not isinstance(steps, int):
-            raise Exception("Please only use integer type for steps and decimal_places with the read_position method")
+            raise ValueError(f"range expects an int value, got: {steps}")
+        percent = self.percent(samples)
+        if percent == 1.0:
+            return steps
+        return int(self.percent(samples) * steps)
+
+    def choice(self, values, samples=None):
+        """Return a value from a list chosen by the relative knob position."""
+        if not isinstance(values, list):
+            raise ValueError(f"choice expects a list, got: {values}")
+        percent = self.percent(samples)
+        if percent == 1.0:
+            return values[-1]
+        return values[int(percent * len(values))]
+
+
+class AnalogueInput(AnalogueReader):
+    def __init__(self, pin, max_voltage=MAX_VOLTAGE, min_voltage=MIN_VOLTAGE):
+        super().__init__(pin)
+        self.MAX_VOLTAGE = max_voltage
+        self.MIN_VOLTAGE = min_voltage
+        self._gradients = []
+        for index, value in enumerate(CALIBRATION_VALUES[:-1]):
+            self._gradients.append(1 / (CALIBRATION_VALUES[index+1] - value))
+        self._gradients.append(self._gradients[-1])
+
+    def percent(self, samples=None):
+        """Current voltage as a relative percentage of the component's range."""
+        # Determine the percent value from the max calibration value.
+        reading = self._sample_adc(samples)
+        max_value = max(reading, CALIBRATION_VALUES[-1])
+        return reading / max_value
+
+    def read_voltage(self, samples=None):
+        reading = self._sample_adc(samples)
+        # low precision vs. high precision
+        if len(self._gradients) == 2:
+            cv = 10 * (reading / CALIBRATION_VALUES[-1])
         else:
-            return int(self.unit_interval(samples) * steps) #Multiply the unit interval by the given steps and convert to integer
+            index = int(self.percent(samples) * (len(CALIBRATION_VALUES) - 1))

```

Comment:
`percent()` will call `_sample_adc()` again. We probably don't want to do that twice in a call to `read_voltage()`

Reply:
Good catch. I've duplicated the `AnalogInput` `percent()` in logic `read_voltage()` to avoid the second call to `_sample_adc()`. Not ideal, but better performance.

Reply:
I think that your recent changes mean that we don't need to duplicate the `percent()` logic anymore.

Reply:
Do you mean the AnalogueInput overridden `percent()` method or the lines calculating `percent` inside the `read_voltage()` method? With high precision calibration, there will be 10 CALIBRATION_VALUES so we do still want to find the correct index from that and `_gradients` to know which calibration value to use when calculating voltage.

Reply:
I mean that the first three lines of `AnalogueInput.read_voltage()` could be replaced with a call to `percent()`.

Reply:
Ah, I see. We still need the raw `reading` in addition to the `percent` value in the voltage calculation. I'm sure there's room for improvement here, so let's add an issue to track it.
```
index = int(percent * (len(CALIBRATION_VALUES) - 1))
cv = index + (self._gradients[index] * (reading - CALIBRATION_VALUES[index]))
```

Reply:
Added #35 to track work with improving the calibration process.